In [1]:
import pandas as pd

In [5]:
disp=pd.read_csv(r"disp.csv",sep=";")
disp.head(5)

,disp_id,client_id,account_id,type
0,1,1,1,OWNER
1,2,2,2,OWNER
2,3,3,2,DISPONENT
3,4,4,3,OWNER
4,5,5,3,DISPONENT


In [6]:
disp.info()

<class 'pandas.DataFrame'>
RangeIndex: 5369 entries, 0 to 5368
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   disp_id     5369 non-null   int64
 1   client_id   5369 non-null   int64
 2   account_id  5369 non-null   int64
 3   type        5369 non-null   str  
dtypes: int64(3), str(1)
memory usage: 167.9 KB


In [7]:
disp["disp_id"] = disp["disp_id"].astype(str)
disp["client_id"] = disp["client_id"].astype(str)
disp["account_id"] = disp["account_id"].astype(str)

In [8]:
disp["type"] = disp["type"].str.upper().str.strip()

## Final Check

In [ ]:
print(disp.info())
print("\nMissing Values:\n", disp.isnull().sum())
print("\nUnique Dispositions:", disp["disp_id"].nunique())
print("\nUnique Clients:", disp["client_id"].nunique())
print("\nUnique Accounts:", disp["account_id"].nunique())
print("\nDuplicates:", disp.duplicated().sum())
print("\nRole Distribution:\n", disp["type"].value_counts())
disp.head()

--- DISP TABLE AUDIT ---
<class 'pandas.DataFrame'>
RangeIndex: 5369 entries, 0 to 5368
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   disp_id     5369 non-null   str  
 1   client_id   5369 non-null   str  
 2   account_id  5369 non-null   str  
 3   type        5369 non-null   str  
dtypes: str(4)
memory usage: 167.9 KB
None

Missing Values:
 disp_id       0
client_id     0
account_id    0
type          0
dtype: int64

Unique Dispositions: 5369

Unique Clients: 5369

Unique Accounts: 4500

Duplicates: 0

Role Distribution:
 type
OWNER        4500
DISPONENT     869
Name: count, dtype: int64


,disp_id,client_id,account_id,type
0,1,1,1,OWNER
1,2,2,2,OWNER
2,3,3,2,DISPONENT
3,4,4,3,OWNER
4,5,5,3,DISPONENT


## Data Cleaning & Transformation — `disp` Table

### 1. Data Cleaning & Transformation Log

| Column Name | Raw Type | Cleaned Type | Original / Raw Value | Converted / Mapped Value | Business / Risk Rationale |
| :--- | :---: | :---: | :--- | :--- | :--- |
| **`disp_id`** | `int64` | `str` | Numeric integer (e.g., `1`) | String object (e.g., `'1'`) | Primary key. Converted to string to prevent numerical math operations and ensure pipeline consistency. |
| **`client_id`** | `int64` | `str` | Numeric integer (e.g., `1`) | String object (e.g., `'1'`) | Foreign key linking to `client.csv`. |
| **`account_id`** | `int64` | `str` | Numeric integer (e.g., `1`) | String object (e.g., `'1'`) | Foreign key linking to `account.csv`. |
| **`type`** | `str` | `str` | String values (`"OWNER"`, `"DISPONENT"`) | Standardized uppercase string (`'OWNER'`, `'DISPONENT'`) | Defines account authorization level. Crucial for filtering primary borrowers (`OWNER`) when linking demographics to loan contracts. |

---

### 2. Validation & Integrity Audit Checklist

- [x] **Null Values:** 0 missing values detected across all columns.
- [x] **Primary Key Uniqueness:** `disp_id` is 100% unique (5,369 distinct records).
- [x] **Role Distribution:** Validated exact mapping (`OWNER`: 4,500, `DISPONENT`: 869). Note that every account (4,500) has exactly 1 `OWNER`.